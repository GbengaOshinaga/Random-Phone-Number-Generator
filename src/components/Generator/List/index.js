import React from 'react'
import { hasReachedBottom, getNumberOfItemsToRender } from '../../../utils'

class List extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      numberToRender: 50
    }
  }

  componentDidMount() {
    document.addEventListener('scroll', this.onEndReached)
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.onEndReached)
  }

  onEndReached = () => {
    const { data } = this.props
    const { numberToRender } = this.state
    const element = document.getElementById('list')
    if (hasReachedBottom(element)) {
      if (data.length > numberToRender) {
        this.setState({ numberToRender: getNumberOfItemsToRender(data.length, numberToRender) })
      }
    }
  }

  renderNumbers = () => {
    const { data } = this.props
    const { numberToRender } = this.state

    return (
      <ul className="list" id="list">
        {data.filter((number, index) => index < numberToRender).map(number => (
          <li key={number} className="item">{number}</li>
        ))}
      </ul>
    )
  }

  render() {
    const { data } = this.props

    return (
      <div>
        {data.length ? this.renderNumbers() : null}
      </div>
    )
  }
}

export default List