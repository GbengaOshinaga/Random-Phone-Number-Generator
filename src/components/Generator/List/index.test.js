import React from 'react'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import List from './index'

it('renders correctly', () => {
  const tree = renderer.create(<List data={[1]} />).toJSON()
  expect(tree).toMatchSnapshot()
})

describe('componentDidMount', () => {
  it('should register listener', () => {
    const map = {}
    document.addEventListener = jest.fn((event, callback) => {
      map[event] = callback
    })
    const list = mount(<List data={[]} />)
    list.instance().onEndReached = jest.fn()
    list.instance().componentDidMount()
    map.scroll({ key: 'scroll' })
    expect(list.instance().onEndReached).toHaveBeenCalled()
  })
})

it('onEndReached', () => {
  const list = mount(<List data={[]} />)
  list.instance().onEndReached()
  expect(list.instance().state.numberToRender).toEqual(50)
})

