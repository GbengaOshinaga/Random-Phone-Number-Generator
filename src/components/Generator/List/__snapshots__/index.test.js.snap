// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <ul
    className="list"
    id="list"
  >
    <li
      className="item"
    >
      1
    </li>
  </ul>
</div>
`;
